File: Solutions/0414- Third Maximum Number/Solution.java
import java.util.HashSet;
import java.util.PriorityQueue;
import java.util.Set;

class Solution {
    public int thirdMax(int[] nums) {
        PriorityQueue<Integer> minHeap = new PriorityQueue<>();
        Set<Integer> set = new HashSet<>();

        for (int num : nums) {
            if (minHeap.size() < 3 && !set.contains(num)) {
                minHeap.add(num);
                set.add(num);
            } else if (num > minHeap.peek() && !set.contains(num)) {
                    minHeap.remove();
                    minHeap.add(num);
                    set.add(num);
                }
        }
        if (minHeap.size() == 3) {
             return minHeap.remove();
        }

        int answer = 0;
        while (!minHeap.isEmpty()) {
            answer = minHeap.remove();
        }
        return answer;

    }
}

File: Solutions/0414- Third Maximum Number/Solution.py
from heapq import heappush, heappop
from typing import List

class Solution:
    def thirdMax(self, nums: List[int]) -> int:
        min_heap = []
        unique_nums = set()

        for num in nums:
            if len(min_heap) < 3 and num not in unique_nums:
                heappush(min_heap, num)
                unique_nums.add(num)
            elif num > min_heap[0] and num not in unique_nums:
                heappop(min_heap)
                heappush(min_heap, num)
                unique_nums.add(num)

        if len(min_heap) == 3:
            return heappop(min_heap)

        return max(min_heap)
        